lib/spack/spack/spec.py

```diff
@@ -667,67 +663,6 @@ def __init__(self, owner):
         # Owner Spec for the current map
         self.owner = owner
 
-    @property
-    def concrete(self):
-
-        # Check if the dependencies are concrete and of the correct type
-        dependencies_are_concrete = all(
-            (d.spec.concrete and d.deptypes) for d in self.values()
-        )
-
-        if not dependencies_are_concrete:
-            return False
-
-        # If we are dealing with an external spec, it clearly has no
-        # dependencies managed by spack
-        if self.owner.external:
-            return True
-
-        # Now check we have every dependency we need
-        pkg = self.owner.package
-        for dependency in pkg.dependencies.values():
-            # Check that for every condition we satisfy we satisfy also
-            # the associated constraint
-            for condition, constraint in dependency.items():
-                # WARNING: This part relies on the fact that dependencies are
-                # the last item to be checked when computing if a spec is
-                # concrete. This means that we are ensured that all variants,
-                # versions, compilers, etc. are there with their final value
-                # when we call 'Spec.satisfies(..., strict=True)'
-
-                # FIXME: at the moment check only for non conditional
-                # FIXME: dependencies, to avoid infinite recursion
-
-                # satisfy_condition = self.owner.satisfies(
-                #     condition, strict=True
-                # )
```

Comment:
I am so glad we removed this function. Fixing it following the old logic would have been a real pain.

Reply:
Fixing it following the old logic would essentially require reimplementing `concretize()` 😄 .  I looked at that function as part of the work on dependency patching and it's also just wrong.  It's the fact that these `concrete()` functions are still called lazily as part of `concretize()` that was screwing everything up -- they weren't really intended to do that.  `Spec` is going to get a lot nicer soon.

Reply:
and yes this function is horrible.